Implement support for Mapview widgets.

maniac103 opened the pull request:
Fixes #574.

maniac103 commented:
See https://github.com/openhab/openhab-android/issues/574#issuecomment-360717159 for screenshot.

maniac103 commented on mobile/src/main/AndroidManifest.xml:
```diff
@@ -114,5 +114,8 @@
                     android:resource="@xml/global_tracker" /> -->
+
+        <meta-data android:name="com.google.android.geo.API_KEY"
+            android:value="AIzaSyBkvcuBAZXKfBZgpfCZji1Qbs49MiXm_Ws" />
```
What's the best way to deal with this? Currently this is my API key, which obviously needs to change. Do we want to add a string resource for it? Or use a gradle variable populated by some (not checked in) gradle resources file?

mueller-ma replied:
From https://developers.google.com/maps/documentation/android-api/signup?hl=en

> or Android apps, you restrict the key to your app's SHA-1 fingerprint and package name.

So I think @digitaldan can get a api key that only works for this app, so we could make it public.

mueller-ma replied:
IIRC the manifest is merged from main + flavor. So please move the api key to mobile/src/full/AndroidManifest.xml

maniac103 replied:
Will do so (along with other full-only stuff, such as the GCM permissions) once I got the API key to use.

mueller-ma replied:
I think you should do this in a new PR.

maniac103 replied:
As you prefer, I'm fine with it either way. I need the API key beforehand nevertheless, though ;-)

maniac103 replied:
> or Android apps, you restrict the key to your app's SHA-1 fingerprint and package name.

Thinking about it, the quoted part means that own builds won't have MapView support though. Maybe we should indeed pull it from a gradle properties file and make the full build fail if it's not defined?

mueller-ma replied:
> Maybe we should indeed pull it from a gradle properties file and make the full build fail if it's not defined?

The gradle file must be encrypted, right? In this case mapview support would also be absend in own builds if you dont have an own api key. This would require for everyone who want to test something with oh cloud (no foss build possible) to get a api key.
I'd rather have no mapview support in own full builds instead of having a failing build.

maniac103 replied:
> The gradle file must be encrypted, right?

No, I would keep it local, that is, not checked in. The Travis environment for doing the Play Store builds would need to have it deployed somehow.

> I'd rather have no mapview support in own full builds instead of having a failing build.

OK. My line of thinking was people not realizing they need to have a local properties file and then wondering why MapView isn't working.

mueller-ma commented:
Are you interessted in implementing mapview in the foss build?

maniac103 commented:
Not at this point at least.

mueller-ma commented on mobile/build.gradle:
```diff
@@ -69,6 +69,7 @@ dependencies {
     compile 'com.android.support:multidex:1.0.2'
     compile 'com.google.code.gson:gson:2.8.0'
     fullCompile 'com.google.android.gms:play-services-gcm:11.4.2'
+    fullCompile 'com.google.android.gms:play-services-maps:11.4.2'
```
11.8.0 is available. Do you want to bump both gms packages here or make a new PR for that?

maniac103 replied:
I'd prefer a new PR for it, I think. I'm not sure whether there are any side effects.

mueller-ma commented:
Looks good so far. Are multiple markers already implemented server side?
I tested readOnly by changing MapViewHelper Line 166 to true/false and that also works.

mueller-ma commented on mobile/src/main/java/org/openhab/habdroid/model/OpenHABItem.java:
```diff
@@ -104,6 +109,10 @@ public void setGroupType(String groupType) {
 		this.groupType = groupType;
 	}
 
+	public boolean isReadOnly() {
```
Can you add a unit test for this?

mueller-ma commented:
Can you extend the expresso tests to check if mapview is there?

maniac103 commented:
> Looks good so far. Are multiple markers already implemented server side?

Nope. See comments in #574.

> Can you extend the expresso tests to check if mapview is there?

Are they always run against the full build?

mueller-ma commented:
> Are they always run against the full build?

They are run with both builds. But there should be a way to check if it runs against full or foss build

maniac103 commented:
If you can tell me how to do that, I'll update the test. I'm a noob in unit testing :-/

mueller-ma commented:
You could execute https://github.com/openhab/openhab-android/pull/686/files#diff-de8b4923c77cb5392822ea907b910f64R9 and check the result.

maniac103 commented:
Unit tests updated.

mueller-ma commented:
Are the maps cached? If not, do you want to implement caching?

maniac103 commented:
> Are the maps cached? If not, do you want to implement caching?

I have no idea what Play Services (in particular: MapView) does internally, but I definitely don't want to fiddle with its internals ;-)

maniac103 commented:
I should probably use lite mode for the in list view though: https://developers.google.com/maps/documentation/android-api/lite

maniac103 commented:
Enabled lite mode for list item and moved API key to a separate properties file ('keys.properties'); updated readme accordingly.

Should be fine now, I think?

mueller-ma commented:
I would keep a restricted api key in the manifest in the repo. See #708

mueller-ma commented:
I would keep a restricted api key in the manifest in the repo. See #708

mueller-ma commented:
I would keep a restricted api key in the manifest in the repo. See #708

maniac103 commented:
In my testing leaving the API key empty did work as well. Not sure whether that was coincidence though, and/or what the limitations of that scenario are.

mueller-ma commented:
The standart free subscription has unlimited free usage. As we dont use things like google places, is there a reason to keep to api key secret?

maniac103 commented:
I'm not too familiar with their pricing plans, so I'm not sure. As long as it's not my key that is added, I'm fine with whatever you prefer.

maniac103 commented:
Updated to fetching the API key property from gradle.properties (as per the example: https://developers.google.com/maps/documentation/android-api/map-with-marker)

mueller-ma commented:
```
:mobile:compileFossDebugAndroidTestJavaWithJavac/home/travis/build/openhab/openhab-android/mobile/src/androidTest/java/org/openhab/habdroid/ui/BasicWidgetTest.java:111: error: cannot find symbol
                    .check(matches(atPositionOnView(13, isDisplayed(), R.id.mapView)));
                                                                           ^
  symbol:   variable mapView
  location: class id
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
```

You could create a dummy mapview layout in the foss flavor. Just something with id=mapView

maniac103 commented:
Fixed it differently (by using View.findViewWithTag).

mueller-ma commented:
@maniac103 Basic ui now shows label and icon: https://github.com/eclipse/smarthome/pull/5071
Can you please update your PR?

maniac103 commented:
Do you mean
- to additionally show the icon only or
- to show icon + label on top or
- to show icon + label in bigger font?

mueller-ma commented:
Like a text item before the mapview, but without value and divider?

maniac103 commented:
So like in https://github.com/openhab/openhab-android/issues/574#issuecomment-360463836 ? Not sure what to think about it :-/ On one hand it's consistent to other items, on the other hand it looks too large in my eye (the map item is a special case as it's the only case that has something below the text)

mueller-ma commented:
> So like in #574 (comment) ?

Yes, but I would move the icon to the left side.

maniac103 commented:
The icon *is* (or rather *would be, if it wasn't empty* ;) ) on the left side.

mueller-ma commented:
Ok, than display this line above the mapview.

maniac103 commented:
Option 1 (consistent with other lines, but ugly IMHO):
![screenshot_1519030157](https://user-images.githubusercontent.com/781373/36369163-81455454-155a-11e8-906b-35760256de62.png)


Option 2 (smaller icon + text):
![screenshot_1519030104](https://user-images.githubusercontent.com/781373/36369177-911aa10e-155a-11e8-8929-f6033d06ca13.png)

Currently implemented is option 2, because I both think it looks better and avoids ambiguities (as option 1 might be mistaken for another widget). If you strongly prefer option 1, I can change it towards that, though.

kaikreuzer commented:
I also think that option 2 looks better.

mueller-ma commented:
Go for 2

maniac103 commented:
Option 2 already is implemented :-)

maniac103 commented:
Removed the marker action toolbar, and added support for the 'height' sitemap parameter. Should be fine now.

maniac103 commented:
Rebased to master. Anything specific holding this back?

mueller-ma commented:
> Anything specific holding this back?

The api key. Dan wants to look at this soon.

maniac103 commented:
Rebased to master.

@digitaldan Any progress with the API key?

BTW, I wonder whether we shouldn't remove gradle.properties from the repository. It only contains commented stuff anyway, and doing so would allow adding the API key without causing a local diff and without having to resort to ~/.gradle/gradle.properties.

mueller-ma commented:
Maybe @kaikreuzer can help us with the api key.
@kaikreuzer can you go to this [site](https://developers.google.com/maps/android/)
* log in
* click on "Schlüssel anfordern"
* create a new project
* click on "restrict this key's usage in the API Console."
* Under "Anwendungseinschränkungen" select Android app
* "Paketname und Fingerabdruck hinzufügen"
* Paketname => org.openhab.habdroid
* SHA1 => `28:E3:A4:1B:7C:84:48:3D:ED:3D:F7:B8:FC:27:89:B5:C9:12:BA:DD`
* Click in "Speichern"
* Now you can post the api key here

mueller-ma commented:
> BTW, I wonder whether we shouldn't remove gradle.properties from the repository. It only contains commented stuff anyway, and doing so would allow adding the API key without causing a local diff and without having to resort to ~/.gradle/gradle.properties.

The api key will be restricted to the signing key and package name, so it can be checked in without problems. After the key is commited, you can add gradle.properties to .gitignore, so local changes won't be tracked.

maniac103 commented:
> After the key is commited, you can add gradle.properties to .gitignore, so local changes won't be tracked.

Adding a committed file to .gitignore? I somehow doubt that's a good idea.

maniac103 commented:
Rebased to master.

mueller-ma commented:
> Adding a committed file to .gitignore? I somehow doubt that's a good idea.

Though it works in this case.

mueller-ma commented:
Where do you want to put our key than? It should be checked in.

maniac103 commented:
IMHO it shouldn't be, or at least does not need to be, as it's useless to anyone building from source anyway.
I'd envision it to be deployed to Jenkins only and to be added to ~/.gradle/gradle.properties there.

mueller-ma commented:
@maniac103 Can you rebase and run `./gradlew -q calculateChecksums` and add the hash for the maps library?

maniac103 commented:
Done. I found the checksums for the GCM library to be missing though, probably because the witness plugin doesn't support flavors.

mueller-ma commented:
One last rebase: Please adjust the string https://github.com/openhab/openhab-android/blob/master/assets/store_descriptions/en-US/strings.sh#L33
I'm going to review this PR and merge it. We can add the key later.

maniac103 commented:
Done.

mueller-ma commented:
@kaikreuzer Can you have a look at my comment above?

kaikreuzer commented:
Sorry, I had missed your previous ping.
Just tried the steps, but on step 3 (create new project), it fails with:

<img width="652" alt="screen shot 2018-03-21 at 16 17 06" src="https://user-images.githubusercontent.com/3244965/37718648-7b4745ea-2d23-11e8-8d6c-a43a87114b79.png">

Will check what I can do about it...

mueller-ma commented:
Did you try a reload without cache? `Controll` + `F5` in FF.

kaikreuzer commented:
Yes. Also tried with Chrome, Safari and FF. It works with my personal Google account, but not with the openHAB Foundation's one... Will try with another openHAB Foundation account, but I don't yet have access to it myself. Stay tuned....

mueller-ma commented:
I dont think it matters which account you use. I just dont want to use my account for two reasons:
1. If I ever leave oh, you still have acces to the key and can change it
2. Maybe it's possible to see which account was used and I want to keep it private

kaikreuzer commented:
> I dont think it matters which account you use.

It does!

> I just dont want to use my account for two reasons

The same reasons apply to my personal account :-)
That's why we create all such accounts under the openHAB Foundation, so that the access to them can be easily transferred to an other foundation member, if people change.

But I have just managed to get access to the account that we also use for google-assistant and I have created the requested Maps API key. Here it is: `AIzaSyA2t2893e9wccs9EnlnauJRgtqFNYQSp-o`

mueller-ma commented:
@maniac103 Please add this to .travis.yml inside the install block:

```
- echo "AIzaSyA2t2893e9wccs9EnlnauJRgtqFNYQSp-o" >> ~/.gradle/gradle.properties
```
